Algorithm.Framework/Risk/MaximumDrawdownPercentPortfolio.cs

```diff
@@ -0,0 +1,108 @@
+
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MaximumDrawdownPercentPortfolio"/> class
+        /// </summary>
+        /// <param name="maximumDrawdownPercent">The maximum percentage drawdown allowed for algorithm portfolio
+        /// compared with starting value, defaults to 5% drawdown</param>
+        /// <param name="email">Optional - Email address for Algorithm Termination Notifications</param>
+        /// <param name="phoneNumber">Optional - Phone Number for Algorithm Termination SMS Notifications</param>
+        public MaximumDrawdownPercentPortfolio(
+            decimal maximumDrawdownPercent = 0.05m, string email = "", string phoneNumber = ""
+            )
+        {
+            _maximumDrawdownPercent = -Math.Abs(maximumDrawdownPercent);
+            _email = email;
+            _phoneNumber = phoneNumber;
+        }
+
+        /// <summary>
+        /// Manages the algorithm's risk at each time step
+        /// </summary>
+        /// <param name="algorithm">The algorithm instance</param>
+        /// <param name="targets">The current portfolio targets to be assessed for risk</param>
+        public override IEnumerable<IPortfolioTarget> ManageRisk(QCAlgorithmFramework algorithm, IPortfolioTarget[] targets)
+        {
+            if(!_initialised)
+            {
+                _startingValue = algorithm.Portfolio.TotalPortfolioValue; // Set inital portfolio value
+                _initialised = true;
+            }
+
+            var pnl = TotalDrawdownPercent(algorithm.Portfolio);
+            if (pnl < _maximumDrawdownPercent)
+            {
+                var message = $"Maximum Portfolio Drawdown Reached - Drawdown: {pnl.SmartRounding()}%, Drawdown Limit: {_maximumDrawdownPercent.SmartRounding()}%";
+                algorithm.Log(message);
+
+                var subject = "Terminating Algorithm";
+                algorithm.Error(subject);
+
+                // Inform user
+                if (_email != "") algorithm.Notify.Email(_email, subject, message);
+                if (_phoneNumber != "") algorithm.Notify.Sms(_phoneNumber, subject + ", " + message);
+
+                KillAlgorithm(algorithm);
+            }
+
+            return targets;
+        }
+
+        private decimal TotalDrawdownPercent(SecurityPortfolioManager portfolio)
+        {
+            var currentValue = portfolio.TotalPortfolioValue;
+            return (currentValue / _startingValue) - 1.0m;            
+        }
+
+        private decimal TotalDrawdownValue(SecurityPortfolioManager portfolio)
+        {
+            return portfolio.TotalProfit + portfolio.TotalUnrealisedProfit;
+        }
+
+        private void KillAlgorithm(QCAlgorithmFramework algorithm)
+        {
+            // Liqudate
+            algorithm.Log($"Liquidating All Holdings");
+            algorithm.Liquidate();
+
+            // Stop algorithm
+            algorithm.Quit("Terminated Algorithm due to Drawdown");
```

Comment:
The framework models in Lean should not terminate the algorithm. I think that liquidating all (or just some) positions should be enough.

Reply:
I can do that but my issues is I don't know what the Order Execution Model will do. It might be expecting to manage all the orders and if I Liquidate in the Risk Model, it might mess up the execution model - so that's why I called Quit, to avoid the Execution Model doing anything unpredictable.

Alternatively I could just change all the Targets to 0 quantity but then it's relying on the Execution Model to do the orders which take some time... 

Which approach would you prefer?

Reply:
The right approach follows the separation of concerns and rely on the execution model to liquidate the positions, so you should change all the targets to zero.
Currently, you are using `Liquidate` method. What if the users disables it?
```csharp
public override void Initialize() {
    Settings.LiquidateEnabled = false;
}
```
and pick an execution model that does not use `Liquidate()`?

Reply:
OK that makes sense, I'll remove it